### DISCUSSION TITLE
More "native order" decoders?

### DISCUSSION THREAD

**Discussion by Zahariel** [python-pillow/Pillow#5237] [Q&A] [answered]
TL;DR: can we add some more "native byte order" raw unpacking modes other than `I:32N`, and/or more "predictable byte order" raw packing modes (which `I;32S` isn't)?

I'm having a problem specifically coming from pycairo. Specifically, pycairo's `FORMAT_ARGB32` format, which is "documented" (I use the term somewhat loosely) as storing each pixel in ARGB, with premultiplied alpha, in "native order". On a little-endian machine (i.e. almost everywhere), this means that the data comes out of `surface.get_data()` in what Pillow would call `BGRa`. But on a big-endian machine, if you can find such a creature, it comes out as `aRGB`. I would like to be able to convert this, portably, to a more predictable format, for example to display in pygame (please just ignore my Rube Goldberg display system here).

I eventually constructed the formulation:
```python
buf = cairo_surface.get_data() # either aRGB or BGRa
pil_image = Image.frombuffer("I", size, bytes(buf), "raw", "I;32B", 0, 1)
fixed_buf = pil_image.tobytes("raw", "I;32S") # implied "native"; definitely aRGB
pygame_image = pygame.image.frombuffer(fixed_buf2, size, "ARGB")
pygame_screen.blit(pygame_image, (0,0), special_flags=pygame.BLEND_PREMULTIPLIED)
```
Which gets the job done, technically, but really isn't clear at all, especially since I really wanted to read `I;32N` and output `I;32B`; but no such packing mode exists, so I swapped them. But what I really, actually wanted, was a notional `aRGB;N` unpacking mode that means "the 32-bit values are aRGB, but they're in machine-native order, so it's probably actually BGRa; you figure it out". In fact, even unpacking `aRGB` as `RGBA` (which is what pygame really wants, although modern versions are willing to put up with premultiplied alpha) doesn't exist; only `BGRa` -> `RGBA` works, which I could do with a 2-step process. It looks quite silly, but it does work.

```python
buf = cairo_surface.get_data() # either aRGB or BGRa
pil_image = Image.frombuffer("I", size, bytes(buf), "raw", "I;32", 0, 1) # now using little-endian
fixed_buf = pil_image.tobytes("raw", "I;32S") # implied "native"; definitely BGRa

pil_image2 = Image.frombuffer("RGBA", size, fixed_buf, "raw", "BGRa", 0, 1)
fixed_buf2 = pil_image2.tobytes("raw", "RGBA")

pygame_image = pygame.image.frombuffer(fixed_buf2, size, "RGBA")
pygame_screen.blit(pygame_image, (0,0))
```

(What I really, truly, actually wanted was for cairo's `get_data()` to do something more predictable, but that ship has most likely sailed.)

So my questions here are: 
1. is there a better way for me to do this? Have I just hacked myself into a ridiculous corner and I should really do this totally differently? 
2. would it be plausible to add my notional `aRGB;N` and similar machine-order-based raw unpacking modes?
3. additionally, you may have noted my use of `bytes(buf)` creating a Pillow image from a pycairo surface. `surface.get_data()` returns a `memoryview`. I would think this is bytes-like enough for Pillow's use? But it doesn't work.

**Comment by radarhere** [comment, marked as answer]
> 1. is there a better way for me to do this? Have I just hacked myself into a ridiculous corner and I should really do this totally differently?

Why isn't https://pycairo.readthedocs.io/en/latest/integration.html#pillow-pil-cairo sufficient for your purposes?

Or, if you're only using Pillow as a stepping stone, https://pycairo.readthedocs.io/en/latest/tutorial/pygame.html#pygame-imagesurface seems like it would solve your problem.

**Reply by Zahariel** [reply]
Your second suggestion is what we had in the first place, and it doesn't actually work (or at least it didn't when I asked this question a couple years ago), no matter how much the documentation says it should; the red and blue channels come out swapped. That's what I was trying to fix.

Your first suggestion is going in the wrong direction. Cairo's *import* utilities are correctly implemented and predictable; it's the *export* from cairo that doesn't work correctly.

**Reply by radarhere** [reply]
Direct to Pygame: This might be in progress with https://github.com/pygame/pygame/pull/3342, coming from https://github.com/pygame/pygame/issues/2972 and https://github.com/pygobject/pycairo/issues/247

**Reply by radarhere** [reply]
The pygame PR has been merged, so directly converting from pycairo to pygame [should now be possible](https://github.com/pygame/pygame/issues/2972#issuecomment-1201951859) with pygame main.

**Reply by Zahariel** [reply]
Nice! I'm not running pygame:main, but hopefully 2.1.3 will be released sometime soon.

**Reply by radarhere** [reply]
pygame 2.1.3 has now been released.

**Comment by radarhere** [comment]
> 3. additionally, you may have noted my use of `bytes(buf)` creating a Pillow image from a pycairo surface. `surface.get_data()` returns a `memoryview`. I would think this is bytes-like enough for Pillow's use? But it doesn't work.

With PR #6974, this should work with Pillow 9.5.0.

**Comment by radarhere** [comment]
> 2. would it be plausible to add my notional aRGB;N and similar machine-order-based raw unpacking modes?

The closest we have at the moment is "RGBa;16N".

However, looking at the code, it becomes clear that the channel order isn't swapped depending on the endianness, simply the values of each channel individually.

https://github.com/python-pillow/Pillow/blob/e7fa3091a065a51ceb7e656b71f67ee587f86082/src/libImaging/Unpack.c#L1619-L1634

https://github.com/python-pillow/Pillow/blob/e7fa3091a065a51ceb7e656b71f67ee587f86082/src/libImaging/Unpack.c#L842-L888

So I'd be reluctant to introduce confusion into the process by introducing this new kind of unpacker.

I think that answers all of your questions. Let us know if not.
